File: tiq-e/dynamic-programming/4/main.py
from time import perf_counter
from typing import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        rob1, rob2 = 0, 0

        for n in nums:
            temp = max(n + rob1, rob2)
            rob1 = rob2
            rob2 = temp
        return rob2


n = [2, 7, 9, 3, 1]

start = perf_counter()
result = Solution().rob(n)
end = perf_counter()
print("Duration: " + str(end - start))
print("Result: " + str(result))

File: tiq-e/dynamic-programming/4/main.rs
use std::time::Instant;

struct Solution;

impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        let mut former_max = 0;
        let mut curr_max = 0;
        for &num in nums.iter() {
            let mut temp = curr_max;
            curr_max = i32::max(former_max + num, curr_max);
            former_max = temp;
        }
        curr_max
    }
}

fn main() {
    let n = Vec::from([2, 7, 9, 3, 1]);

    let start = Instant::now();
    let result = Solution::rob(n);
    let end = start.elapsed().as_micros();
    println!("Duration: {:?}", end);
    println!("Result: {:?}", result);
}
